### PR TITLE
util: move codec logic into tablecodec package

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What is changed and how it works?

Proposal: #17932 

What's Changed:
Try to move the index codec related logic into `tablecodec`  package so that they can be reused. Encoded index keys and values can be generated with known `model.TableInfo` and `model.IndexInfo` from the schema meta.

How it Works:

### Check List 

Tests 

- Unit test

Side effects

### Release note 

-

### CODE REVIEW

File: tablecodec/tablecodec.go [hunk 1]
```diff
@@ -934,3 +938,190 @@ func GetTableIndexKeyRange(tableID, indexID int64) (startKey, endKey []byte) {
...
+func GetIndexKeyBuf(buf []byte, defaultCap int) []byte {
+	if buf != nil {
+		return buf[:0]
+	}
+	return make([]byte, 0, defaultCap)
+}
+
+// GenIndexKey creates index keys using input buf
+func GenIndexKey(sc *stmtctx.StatementContext, tblInfo *model.TableInfo, idxInfo *model.IndexInfo,
+	idxPrefix kv.Key, indexedValues []types.Datum, h kv.Handle, buf []byte) (key []byte, distinct bool, err error) {
```
Comment (coocood) [inline tablecodec/tablecodec.go:952]: Can we remove `idxPrefix`?

Comment (cfzjywxk) [reply tablecodec/tablecodec.go:952]: The table id could be `physicalTableID` which is different from `TableInfo.ID` for partition table, `prefix` or `physicalTableID` is still needed as param.

Comment (coocood) [reply tablecodec/tablecodec.go:952]: I think `physTblID` is better, the caller don't need to create idxPrefix.
In the amend index case, the `physTblID` can be extracted from the row key.

Comment (cfzjywxk) [reply tablecodec/tablecodec.go:952]: Got

Comment (cfzjywxk) [reply tablecodec/tablecodec.go:960]: A new method "GenIndexKeyUsingPrefix" is wrapped, `Index.Create` will use this method with the already encoded prefix byte array.  "GenIndexKey" has physical table id as input param and will do the prefix encode itself.

File: tablecodec/tablecodec.go [hunk 2]
```diff
@@ -934,3 +938,190 @@ func GetTableIndexKeyRange(tableID, indexID int64) (startKey, endKey []byte) {
...
+		if h.IsInt() {
+			key, err = codec.EncodeKey(sc, key, types.NewDatum(h.IntValue()))
+		} else {
+			key = append(key, h.Encoded()...)
+		}
+	}
+	return
+}
+
+// GenIndexValue creates encoded index value and return the result
```
Comment (zimulala) [inline tablecodec/tablecodec.go:986]:
```suggestion
// GenIndexValue creates encoded index value and returns the result.
```

**Comment by cfzjywxk** [comment]
@zimulala @wjhuang2016 need help

**Pull request merged**